packages/liveblocks-core/src/ai.ts

```diff
@@ -952,6 +952,42 @@ export function createAi(config: AiConfig): Ai {
     knowledge: new KnowledgeStack(),
   };
 
+  // Delta batch processing system to throttle incoming delta updates. Incoming
+  // deltas are buffered and only let through every every 25ms. This creates
+  // a ceiling of max 40 rerenders/second during streaming.
+  const DELTA_THROTTLE = 25;
+  let pendingDeltas: { id: MessageId; delta: AiAssistantDeltaUpdate }[] = [];
+  let deltaBatchTimer: NodeJS.Timeout | number | null = null;
+
+  function flushPendingDeltas() {
+    const currentQueue = pendingDeltas;
+
+    pendingDeltas = [];
+    if (deltaBatchTimer !== null) {
+      clearTimeout(deltaBatchTimer);
+      deltaBatchTimer = null;
+    }
+
+    // Process all pending deltas in a single batch
+    batch(() => {
+      console.warn(
+        `Processing ${currentQueue.length} pending deltas in this chunk`
+      );
+      for (const { id, delta } of currentQueue) {
+        context.messagesStore.addDelta(id, delta);
+      }
+    });
+  }
+
+  function enqueueDelta(id: MessageId, delta: AiAssistantDeltaUpdate) {
+    pendingDeltas.push({ id, delta });
+
+    // If no timer is running, start one to process the batch
+    if (deltaBatchTimer === null) {
+      deltaBatchTimer = setTimeout(flushPendingDeltas, DELTA_THROTTLE);
```

Comment:
Should we also `clearTimeout` this when WS disconnects (for any reason, including having no chats or `LiveblocksProvider` rendered anymore)?